### PR TITLE
Fix various issues with System.Json.JsonValue.Save

### LINKED ISSUE #26486: The `System.Json.JsonObject.Save(System.IO.Stream)` method produces invalid json.
The "json" this method writes to the stream puts a comma between keys and values, and does not separate the next key from the previous value with a comma. E.g. when you expect the output `{"foo":"bar","baz":"frop"}`, you will actually get something like `{"foo","bar""baz","frop"}`. This bug has existed since the `JsonObject` class was invented in 2008, see [here](https://github.com/mono/mono/commit/66ecab0058cbf5c8aaa1c9ca89da52752d465eb4#diff-6a07ba1015d9ca7fc1b5b5a10859b17eR162). Perhaps the best fix would be to remove the override of `JsonValue.Save(Stream)` in `JsonObject` entirely; clearly nobody is using it (or this bug would have been noticed before), and the base implementation on `JsonValue` actually does the right thing.

### PR DESCRIPTION
**Pull request by Thiez**
The `JsonObject` class overrides `JsonValue.Save(Stream)`, but was
implemented incorrectly: it would produce output resembling
`{"key1","value1""key2","value2"}`, instead of the more reasonable
`{"key1": "value1", "key2": "value2"}`. Because `JsonValue` already
contains code that can properly serialize a `JsonObject`, the override
now simply calls the base method.

In addition, it turns out that the `JsonValue.Save(Stream)` method was
also implemented incorrectly. This method would invoke
`JsonValue.Save(TetxtWriter)`, but the `StreamWriter` it created was never
flushed or disposed, so output could go missing. In addition, it would
encode json as utf-8, but would include a BOM, while none of the methods
that override `JsonValue.Save(Stream)` would write a BOM. Both these
issues have been resolved.

The tests have also been modified to test for the new and correct
behavior.

Fixes #26486

### CODE REVIEW

File: src/System.Json/src/System/Json/JsonValue.cs [hunk 1]
```diff
@@ -122,7 +122,10 @@ public virtual void Save(Stream stream)
                 throw new ArgumentNullException(nameof(stream));
             }
 
-            Save(new StreamWriter(stream));
+            using (StreamWriter writer = new StreamWriter(stream, new UTF8Encoding(false), 1024, true))
```
[the remark this answers is not in the document]
Comment (stephentoub) [reply src/System.Json/src/System/Json/JsonValue.cs]: > I can however call new UTF8Encoding(false, true) to ensure we throw on invalid bytes.

Right.

> I suppose the UTF8Encoding instance could be stored in a private static readonly field so we don't have to keep creating new instances

Exactly

**Comment by tarekgh** [comment]
@Thiez  do you think this change can break any app used to depend on the old format? or we already have the base class returning the new format anyway so the apps already seeing the new format in other places?

**Comment by tarekgh** [comment]
test Packaging All Configurations x64 Debug Build please

**Comment by tarekgh** [comment]
```
12:13:29 System\Json\JsonValue.cs(18,33): error CS0246: The type or namespace name 'UTF8Encoding' could not be found (are you missing a using directive or an assembly reference?) [D:\j\workspace\windows-TGrou---0d2c9ac4\src\System.Json\src\System.Json.csproj]
```

you may try to move the line 
```
    <Reference Include="System.Text.Encoding" />
```
to the unconditioned ItemGroup block in the csproj file.

**Comment by Thiez** [comment]
> @Thiez do you think this change can break any app used to depend on the old format?

Yes, it surely would. However, given that the old format is so *very* wrong, I doubt there is any app out there that is actually using it in its current form, especially since the bug doesn't manifest when you call `Save(TextWriter)`.

> or we already have the base class returning the new format anyway so the apps already seeing the new format in other places?

The new format of `Save(Stream)` for `JsonObject` is now identical to what calling `Save(TextWriter)` has produced all along. The base `Save(Stream)` method on abstract class `JsonValue` was never called, because all classes that extend `JsonValue` override this method. Perhaps someone out there has created their own overrides of `JsonValue` that uses the base implementation of `Save(Stream)`, but I strongly doubt that, because this method fails to flush its `StreamWriter` and therefor fails to produce any output most of the time.

Thanks for the advice on the csproj, I had no idea where to start fixing that. I'll try and see if that works.

**Comment by Thiez** [comment]
@tarekgh I've modified the csproj file as you have suggested, but it doesn't help. I get the following error message:

```
Running: C:\Users\Username\Desktop\test\corefx\Tools\msbuild.cmd /nologo /verbosity:minimal /clp:Summary /maxcpucount /nodeReuse:false /l:BinClashLogger,Tools\net46\Microsoft.DotNet.Build.Tasks.dll;LogFile=binclash.log /p:ConfigurationGroup=Debug  /flp:v=normal  /flp2:warningsonly;logfile=msbuild.wrn  /flp3:errorsonly;logfile=msbuild.err  src/dirs.proj /p:DirectoryToBuild=C:\Users\Username\Desktop\test\corefx\src\System.Json

  System.Json -> C:\Users\Username\Desktop\test\corefx\bin\AnyOS.AnyCPU.Debug\System.Json\netstandard\System.Json.dll
  System.Json.Tests -> C:\Users\Username\Desktop\test\corefx\bin\AnyOS.AnyCPU.Debug\System.Json.Tests\netcoreapp\System.Json.Tests.dll
  Using C:\Users\Username\Desktop\test\corefx\bin\testhost\netcoreapp-Windows_NT-Debug-x64\ as the test runtime folder.
  Executing in C:\Users\Username\Desktop\test\corefx\bin\tests\System.Json.Tests\netcoreapp-Windows_NT-Debug-x64\
  Running tests... Start time:  0:09:59.34
  Error:
    An assembly specified in the application dependencies manifest (Microsoft.NETCore.App.deps.json) was not found:
      package: 'runtime.win-x64.Microsoft.NETCore.App', version: '2.0.0'
      path: 'runtimes/win-x64/lib/netcoreapp2.0/System.Runtime.Caching.dll'
  Finished running tests.  End time= 0:09:59.43, Exit code = -2147450740 
```

**Comment by tarekgh** [comment]
@Thiez I'll take a look and will get back to you.

**Comment by tarekgh** [comment]
please add

```Xml
    <Reference Include="System.Text.Encoding" />
```
after the line 

```Xml
    <Reference Include="System.Text.Encoding.Extensions" />
```

in the original csproj. this should solve the problem.

File: src/System.Json/src/System/Json/JsonValue.cs [hunk 2]
```diff
@@ -122,7 +124,10 @@ public virtual void Save(Stream stream)
                 throw new ArgumentNullException(nameof(stream));
             }
 
-            Save(new StreamWriter(stream));
+            using (StreamWriter writer = new StreamWriter(stream, s_encoding, 1024, true))
```
Comment (tarekgh) [inline src/System.Json/src/System/Json/JsonValue.cs]: why do you pass the last parameter as true here? shouldn't pass it at all or pass it false?

**Comment by tarekgh** [comment]
@pjanotti we have a minor breaking change here. if we need to document this in https://github.com/dotnet/platform-compat, how we can do that? any guideline is appreciated. thanks.

File: src/System.Json/src/System/Json/JsonValue.cs (hunk 2 above)
Comment (Thiez) [reply src/System.Json/src/System/Json/JsonValue.cs]: I'm not sure I understand the question, why should calling this method close the stream? It doesn't mention doing so in the documentation, nor do any of its overriding methods close it.

Comment (tarekgh) [reply src/System.Json/src/System/Json/JsonValue.cs]: the last parameter of ```StreamWriter(stream, s_encoding, 1024, true) ``` is documented as 

```
leaveOpen
Type: System.Boolean
true to leave the stream open after the StreamWriter object is disposed; otherwise, false.

https://msdn.microsoft.com/en-us/library/gg712853(v=vs.110).aspx
```

and you are passing true value. so my question is, why you need to keep the stream opened after disposing of StreamWriterObject? 
Also, the original code used ```StreamWriter(stream)``` which eventually will call the other constructor with parameters like ```C# StreamWriter(stream, s_encoding, 1024, false)```. note that it is passing false to force closing the stream.

Comment (Thiez) [reply src/System.Json/src/System/Json/JsonValue.cs]: The `StreamWriter` in the original code was never disposed, so it made no difference that it defaulted to closing the stream. It also never flushed, meaning only part of the output ended up in the stream (or none of it, if your json object was small enough...).

Comment (tarekgh) [reply src/System.Json/src/System/Json/JsonValue.cs]: > The StreamWriter in the original code was never disposed

Is this true? wouldn't at some point the GC will kick in and finalize this object?

**Comment by pjanotti** [comment]
While sometimes code takes dependency with some odd behavior or bug in this case it simply was unusable as JSON. Perhaps what we want here is to straight document that in 2.1 it was fixed to return actual JSON.

/cc @mairaw @OliaG

**Comment by mairaw** [comment]
@rpetrusha any guidance here on how to document this breaking change?

File: src/System.Json/src/System/Json/JsonValue.cs (hunk 2 above)
Comment (Thiez) [reply src/System.Json/src/System/Json/JsonValue.cs]: First of all neither `StreamWriter` nor its base class `TextWriter` has a finalizer. Which is correct; neither of these classes owns unmanaged resources. The GC doesn't see `IDisposable.Dispose` as special, and will not call it for you.

Secondly even if `StreamWriter` had a finalizer which would call `Dispose`, there is no guarantee that the GC *ever* runs (and, more reasonably, there is no guarantee that the GC would run before you start performing actions that rely on the json-data being written to the stream), so we wouldn't be able to rely on this behaviour anyway.

Thirdly, if `JsonValue.Save(Stream)` would close the stream, it would be impossible to use for most json values. For example, `new JsonArray(new JsonObject(), new JsonObject()).Save(new MemoryStream())` would throw an exception (either `NotSupportedException` or `ObjectDisposedException`, I predict), because `JsonArray` calls `JsonValue.Save(stream)` on its elements, so serializing the first `JsonObject` would close the stream and dispose the stream,

In short, I think we really need to pass `true` as the final constructor parameter :)

Comment (tarekgh) [reply src/System.Json/src/System/Json/JsonValue.cs]: ok. make sense then. thanks for the clarification.

**Comment by tarekgh** [comment]
test OSX x64 Debug Build please
test Linux x64 Release Build please

**Comment by tarekgh** [comment]
@dotnet-bot test this please

**Comment by danmosemsft** [comment]
IMO it seems to me if it was just broken before then nobody was using it so it probably isn't necessary to document that it now works..
